File: src/_110/kotlin/Solution.kt
package _110.kotlin

import structure.TreeNode

/**
 * @author relish
 * @since 2018/04/30
 * <p>
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int = 0) {
 * var left: TreeNode? = null
 * var right: TreeNode? = null
 * }
 */
class Solution {
    fun isBalanced(root: TreeNode?): Boolean {
        if (root == null) return true
        val left = depth(root.left)
        val right = depth(root.right)
        return Math.abs(left - right) <= 1 && isBalanced(root.left) && isBalanced(root.right)
    }

    private fun depth(node: TreeNode?): Int {
        return if (node == null) 0 else 1 + Math.max(depth(node.left), depth(node.right))
    }
}

fun main(args: Array<String>) {
    val node = TreeNode.createTestData("[3,9,20,null,null,15,7]")
//    TreeNode.print(node)
    println(Solution().isBalanced(node))
}

File: src/_110/Solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isBalanced = function(root) {
  if(root == null) return true
  var treeDeep = function(root) {
      if(root == null) return 0
      return 1 + Math.max(treeDeep(root.left),treeDeep(root.right))
  }
  leftDeep = treeDeep(root.left)
  rightDeep = treeDeep(root.right)
  if(Math.abs(leftDeep - rightDeep) > 1) {
      return false   
  }
  return isBalanced(root.left) && isBalanced(root.right)
};